subsys/net/lib/lwm2m/lwm2m_engine.h

```diff
@@ -30,6 +31,36 @@
 /* length of time in milliseconds to wait for buffer allocations */
 #define BUF_ALLOC_TIMEOUT K_SECONDS(1)
 
+/** initialization function */
+struct lwm2m_init_func {
+	int (*f)(void);
+};
+/**
+ * @defgroup LWM2M_PRIO LwM2M initialization priorities
+ * @{
+ */
+#define LWM2M_PRIO_ENGINE 0	/**< Engine initialization */
+#define LWM2M_PRIO_CORE 1	/**< Core object initialization */
+#define LWM2M_PRIO_OBJ 2	/**< Object initializations */
+#define LwM2M_PRIO_APP 3	/**< Application logic initialization */
+/** @} */
+
+/**
+ * @brief Declare an initialization function to be executed when LwM2M engine starts.
+ *
+ * When LwM2M engine starts up, it first executes all initialization functions in following
+ * priority order:
+ * 1. LWM2M_PRIO_ENGINE
+ * 2. LWM2M_PRIO_CORE, this is where all LwM2M core objects are initialized
+ * 3. LWM2M_PRIO_OBJ, this is where all other than core objects are initialized
+ * 4. LwM2M_PRIO_APP, application initialization.
+ *                    For example create sensor objects, and register object callbacks.
+ *
+ * @param[in] prio Priority, one of @ref LWM2M_PRIO macros.
+ */
+#define LWM2M_ON_INIT(prio, init_function)                                                         \
+	STRUCT_SECTION_ITERABLE(lwm2m_init_func,                                                   \
+				CONCAT(LWM2M, prio, init_function)) = {.f = init_function}
```

Comment:
So one macro for each of the level?

* LWM2M_ENG_INIT()
* LWM2M_CORE_INIT()
* LWM2M_OBJ_INIT()
* LWM2M_APP_INIT()

At least it would be shorter.. and as it is macro, it does cost anything..